src/ifcgeom/mapping/IfcCurveSegment.cpp

```diff
@@ -249,48 +206,29 @@ class curve_segment_evaluator {
 		auto fn_x = [theta](double t)->double {return cos(theta(t)); };
 		auto fn_y = [theta](double t)->double {return sin(theta(t)); };
 
-		set_spiral_functor(s->as<IfcSchema::IfcSpiral>(), sign_x, fn_x, sign_y, fn_y);
+		set_spiral_functor(mapping_,s->as<IfcSchema::IfcSpiral>(), sign_x, fn_x, sign_y, fn_y);
 	}
 #endif
 
 	void operator()(IfcSchema::IfcCircle* c)
 	{
 		auto R = c->Radius();
+		//const auto& transformation_matrix = taxonomy::cast<taxonomy::matrix4>(mapping_->map(c->Position()))->ccomponents();
+		auto transformation_matrix = taxonomy::cast<taxonomy::matrix4>(mapping_->map(c->Position()))->ccomponents();
 
-		auto position = c->Position();
-		auto placement = position->as<IfcSchema::IfcAxis2Placement2D>();
-		auto ref_direction = placement->RefDirection();
-		double theta = 0.0; // angle the circle's placement X-axis makes with respect to global X axis
-		if (ref_direction)
-			{
-			auto dr = ref_direction->DirectionRatios();
-			auto dx = dr[0];
-			auto dy = dr[1];
-			theta = atan2(dy, dx);
-		}
-
-		// center of circle location
-		auto C = placement->Location();
-		if (!C->as<IfcSchema::IfcCartesianPoint>())
-		{
-			throw std::runtime_error("Only IfcCartesianPoint is supported for center of IfcCircle");
-			// @todo add support for other IfcPoint subtypes
-		}
-		auto Cx = C->as<IfcSchema::IfcCartesianPoint>()->Coordinates()[0];
-		auto Cy = C->as<IfcSchema::IfcCartesianPoint>()->Coordinates()[1];
-
-		eval_ = [R, Cx, Cy, theta](double u)
+		eval_ = [R, transformation_matrix](double u)
 			{
 				auto angle = u / R; // angle subtended by arc length u
 
 				// compute point on circle centered at (0,0) with x-axis horizontal and y-axis vertical
-				auto xl = R * cos(angle);
-				auto yl = R * sin(angle);
+				auto x = R * cos(angle);
+				auto y = R * sin(angle);
 
 				// transform point into circle's coodinate system
-				auto x = xl * cos(theta) - yl * sin(theta) + Cx;
-				auto y = xl * sin(theta) + yl * cos(theta) + Cy;
-				return Eigen::Vector3d(x, y, 0.0);
+				auto result = transformation_matrix * Eigen::Vector4d(x, y, 0.0, 1.0);
```

Comment:
@aothms is there a better way to apply the transformation matrix? When I multiply matrix4 with vectorxd, the results are not correct

Reply:
`Eigen::Vector3d(x, y, 0.0).homogeneous();` can be used instead of Vector4D. For example when you have a contiguous matrix of XYZ values, homogenous can be used to colwise add the 1 symbollically without making a copy (using Eigen::map)  of the underlying data. I can't find anything on VectorXd in the docs, other than that there would be a runtime check instead of a compile time check on size. I have no problem with temporarily casting to Vector4d. I assume the compiler will help us out here with optimizations.